### PR TITLE
Experiments with mocks for SDK and SwiftUI

### PR DESCRIPTION
**Pull request by maratal**
Research for #4

### CODE REVIEW

File: Example/AblyChatExample/Mocks/Mocks.swift [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    func detach() async throws {
+        fatalError("Not yet implemented")
+    }
+    
+    var options: RoomOptions {
+        fatalError("Not yet implemented")
+    }
+}
+
+final class MockMessages: @unchecked Sendable, Messages, ObservableObject {
```
Comment (lawrence-forooghian) [inline Example/AblyChatExample/Mocks/Mocks.swift:96]: Why the `@unchecked`? That’s used to tell the compiler "trust me, I’ve made sure that this is `Sendable`, you don't need to check it" and normally implies some synchronisation mechanism of your own. It doesn't seem like it should be necessary here.

File: Example/AblyChatExample/ContentView.swift [hunk 2]
```diff
@@ -3,22 +3,130 @@ import SwiftUI
...
+//        realtime: MockRealtime(key: ""),
+//        clientOptions: ClientOptions()
+//    )
+    
+    var room: Room {
+        try! MockChatClient.shared.rooms.get(roomID: "Demo Room", options: .init())
+    }
+    
+    // compiler doesn't allow to use `room` property here, needs to think about proper initialization
+    @StateObject private var messagesClient = try! MockChatClient.shared.rooms.get(roomID: "Demo Room", options: .init()).messages as! MockMessages
```
Comment (lawrence-forooghian) [inline Example/AblyChatExample/ContentView.swift:16]: Why the `as! MockMessages`? This means that we’re no longer just using the API of the SDK (i.e. the `Messages` protocol).

File: Example/AblyChatExample/Mocks/Mocks.swift [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    let clientID: String
+    let roomID: String
+    
+    @Published var log = [Message]()
```
Comment (lawrence-forooghian) [inline Example/AblyChatExample/Mocks/Mocks.swift:100]: What's the purpose of this and the `ObservableObject` conformance? It doesn't seem related to the SDK’s API.

**Comment by lawrence-forooghian** [comment]
(I know that this is a WIP but I’ve put some comments in here since I took a look after you mentioned it in #9.)

File: Example/AblyChatExample/Mocks/Mocks.swift (hunk 1 above)
Comment (maratal) [reply Example/AblyChatExample/Mocks/Mocks.swift:96]: It forced by compiler. Without `@unchecked` it gives an error:
```
Stored property '_log' of 'Sendable'-conforming class 'MockMessages' is mutable
```
If I change `MockMessages` to be a struct, then it's impossible to use it with `ObservableObject` protocol, because it's a class protocol.

File: Example/AblyChatExample/ContentView.swift (hunk 2 above)
Comment (maratal) [reply Example/AblyChatExample/ContentView.swift:16]: This is also forced by a compiler:
```
Type 'any Messages' cannot conform to 'ObservableObject'
```
But this is an experimental PR in which I'm just trying things. An expanded answer is in the https://github.com/ably-labs/ably-chat-swift/pull/9

**Comment by maratal** [comment]
>What do you mean when you say that "you usually don't iterate through it"? In https://github.com/ably-labs/ably-chat-swift/pull/16 you have a for await loop in which you iterate over a subscription.

>Do you mean the current subscribe() method? But that method doesn't fetch anything; it just returns a value which:

I've meant that subscriptions are not iterable by design, they are event-based. They don't have an internal array to iterate through. Yes, I do iterate it in for await loop, but it's something I've found strange to be doing. But maybe it's just me.
So, maybe if

>1. implements AsyncSequence to allow you to receive new messages as they come in

just give it the most obvious name `asyncSequence`? `for await message in room.messages.asyncSequence()`

>2. provides a getPreviousMessages method to allow you to fetch earlier messages

Shouldn't it return the same as `Messages.get(options: QueryOptions)` with `backwards` option?

>I don't really understand what you're doing with that; are you saying that you think that instead of async sequences, the SDK should use stateful objects that are marked with Observable? I wouldn’t recommend that since it'd only work for SwiftUI users.

Yes, you're right, it's SwiftUI only. Playing with it I forgot that the adoption rate of it is still low. Thanks for https://github.com/ably-labs/ably-chat-swift/issues/17

>Are you referring to the fact that ContentView reads the log property which lists messages that have been sent via send? Yeah, I think we need to get rid of that and use the SDK’s API.

The idea was that `log` contains all the messages created via: downloaded history + subscription + send, since you need to display all of them in one place. On send you should display the message immediately even if you don't have the internet. That's what all modern messengers do at least.

>I don't know enough about AsyncSequence (and Sequence which has the same iterator model) to say for sure, but I don't think it's common for a sequence to be its own iterator (because an iterator represents a single iteration over a sequence, but a sequence can in theory be iterated over multiple times). The nested struct is copied from Swift’s AsyncStream.

Will look into that.

File: Example/AblyChatExample/Mocks/Mocks.swift (hunk 3 above)
Comment (maratal) [reply Example/AblyChatExample/Mocks/Mocks.swift:100]: See https://github.com/ably-labs/ably-chat-swift/pull/16#issuecomment-2303070970

File: Example/AblyChatExample/Mocks/Mocks.swift (hunk 1 above)
Comment (lawrence-forooghian) [reply Example/AblyChatExample/Mocks/Mocks.swift:96]: But the compiler is warning you about a genuine data race; suppressing it by doing `@unchecked Sendable` whilst not addressing the data race isn't the right thing to do.

**Comment by lawrence-forooghian** [comment]
> The idea was that log contains all the messages created via: downloaded history + subscription + send, since you need to display all of them in one place. On send you should display the message immediately even if you don't have the internet. That's what all modern messengers do at least (with re-send button for each failed message).

But this does not currently fall within the functionality that the SDK offers, so why are you trying to add this functionality into the `MockMessages` class, which is meant to be a fake version of the SDK? Seems like that's something you should be handling elsewhere in the app.

**Comment by lawrence-forooghian** [comment]
As mentioned the other day, I think it would be best if in this PR we simply created a mock implementation of the protocols, without trying to build an example app around them.

**Comment by maratal** [comment]
I'm closing this in favor of https://github.com/ably-labs/ably-chat-swift/pull/34

**Comment by maratal** [comment]
I'm closing this in favor of https://github.com/ably-labs/ably-chat-swift/pull/34
